File: spring-boot-template/src/main/java/tui/meta/challenge/quotes/controller/QuoteController.java
package tui.meta.challenge.quotes.controller;

import jakarta.validation.Valid;
import jakarta.validation.constraints.Size;
import lombok.RequiredArgsConstructor;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import tui.meta.challenge.quotes.controller.model.QuoteDto;
import tui.meta.challenge.quotes.service.QuoteService;

import java.util.List;

@RestController
@RequestMapping("/api/v1/quotes")
@RequiredArgsConstructor
@Validated
public class QuoteController {

    private final QuoteService quoteService;

    @GetMapping(path = "/{id}", produces = "application/json")
    public QuoteDto getQuote(@PathVariable(value = "id") String id) {
        return quoteService.getSingleQuote(id);
    }

    @GetMapping(produces = "application/json")
    public List<QuoteDto> getQuotesWithSearch(
            @RequestParam(value = "author", required = false)
            @Valid
            @Size(min = 2, max = 100, message = "Author name must be between 2 and 100 characters.")
            String author) {
        if (author != null) {
            return quoteService.getQuotesByAuthor(author);
        }

        return quoteService.getAllQuotes();
    }
}

File: micronaut-template/src/main/kotlin/tui/meta/challenge/controller/QuoteController.kt
package tui.meta.challenge.controller

import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.http.annotation.QueryValue
import jakarta.validation.constraints.Size
import tui.meta.challenge.controller.model.QuoteDto
import tui.meta.challenge.service.QuoteService

@Controller(value = "/api/v1/quotes")
class QuoteController(
    private val quoteServiceImpl: QuoteService
) {

    @Get(uri = "/{id}", produces = [MediaType.APPLICATION_JSON])
    fun getQuote(@PathVariable id: String): QuoteDto = quoteServiceImpl.getQuoteById(id)

    @Get(produces = [MediaType.APPLICATION_JSON])
    fun getQuotesWithSearch(
        @QueryValue("author") @Size(
            min = 2,
            max = 100,
            message = "Author name must be between 2 and 100 characters."
        ) author: String?
    ): List<QuoteDto> {
        if (author != null) {
            return quoteServiceImpl.getQuotesByAuthor(author)
        }

        return quoteServiceImpl.getAllQuotes()
    }

}